### PR TITLE
feat: separate deposit execution logic

### PR DESCRIPTION
**Pull request by nmlinaric**
## Description
This PR separates the following logic:
- deposit logic from Bridge contract to Router contract
- execution logic from Bridge contract to Executor contract 
- Bridge now only has logic for configuring resource/routes,....

## Related Issue Or Context
Closes: #2 

## How Has This Been Tested? Testing details.
- unit tests
- e2e tests

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [x] I have updated the documentation locally and in chainbridge-docs.
- [x] I have added tests to cover my changes.
- [x] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

File: src/contracts/handlers/FeeHandlerRouter.sol [hunk 1]
```diff
@@ -24,13 +24,13 @@ contract FeeHandlerRouter is IFeeHandler, AccessControl {
...
-        _onlyBridge();
+    modifier onlyRouter() {
+        _onlyRouter();
         _;
     }
 
-    function _onlyBridge() private view {
-        require(msg.sender == _bridgeAddress, "sender must be bridge contract");
+    function _onlyRouter() private view {
+        require(msg.sender == _bridgeAddress, "sender must be router contract");
```
Comment (lastperson) [inline src/contracts/handlers/FeeHandlerRouter.sol:33]: Perhaps need to rename the _bridgeAddress into _routerAddress ?

File: src/contracts/handlers/PermissionlessGenericHandler.sol [hunk 2]
```diff
@@ -23,11 +24,21 @@ contract PermissionlessGenericHandler is IHandler {
...
+        _;
+    }
+
+    function _onlyExecutor() private view {
+        require(msg.sender == _executorAddress, "sender must be executor contract");
+    }
+
     /**
-        @param bridgeAddress Contract address of previously deployed Bridge.
+        @param executorAddress Contract address of previously deployed Bridge.
```
Comment (lastperson) [inline src/contracts/handlers/PermissionlessGenericHandler.sol:37]: bridgeAddress still requires a comment.

File: src/contracts/Router.sol [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        - ERC20Handler: responds with an empty data.
+        - PermissionlessGenericHandler: responds with an empty data.
+     */
+    function deposit(
+        uint8 destinationDomainID,
+        bytes32 resourceID,
+        bytes calldata depositData,
+        bytes calldata feeData
+    ) external payable whenBridgeNotPaused returns (uint64 depositNonce, bytes memory handlerResponse) {
+        uint8 domainID = IBridge(_bridge)._domainID();
```
Comment (lastperson) [inline src/contracts/Router.sol:93]: _domainID could be cached in the constructor.

File: src/contracts/Router.sol [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+/**
+    @title Facilitates proposal deposits.
+    @author ChainSafe Systems.
+ */
+contract Router is Context {
+    using ECDSA for bytes32;
+
+    IBridge public immutable _bridge;
+    IAccessControlSegregator public _accessControl;
```
Comment (lastperson) [inline src/contracts/Router.sol:24]: Should there be multiple access controls? Separate for Bridge and Router? Or maybe a single one on Bridge is enough?

File: src/contracts/Executor.sol [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        @notice If executeProposals fails it doesn't revert, emits {FailedHandlerExecution} event.
+        @param proposals Array of Proposal which consists of:
+        - originDomainID ID of chain deposit originated from.
+        - resourceID ResourceID to be used when making deposits.
+        - depositNonce ID of deposit generated by origin Bridge contract.
+        - data Data originally provided when deposit was made.
+        @notice Emits {ProposalExecution} event for each proposal in the batch.
+        @notice Behaviour: when execution fails, the handler will terminate the function with revert.
+     */
+    function executeProposals(Proposal[] memory proposals) public whenBridgeNotPaused {
```
Comment (lastperson) [inline src/contracts/Executor.sol:59]: Access control? There should be a test for this.

**Review by lastperson — CHANGES_REQUESTED** [review]

File: src/contracts/Executor.sol (hunk 5 above)
Comment (lastperson) [reply src/contracts/Executor.sol:59]: I mean, executeProposal can be called by anyone, which does not seem correct.

Comment (nmlinaric) [reply src/contracts/Executor.sol:59]: Yes, yes, should've be more descriptive about this. It looks a bit weird right now, this is a middle step and in the following PR I'll add proof verification logic to executeProposal, where on source chain we'll store a hash of all the data in the deposit and on the destination chain we'll check if that has been included in the block state.

**Review by tcar121293 — APPROVED** [review]

**Pull request merged**